src/v/coproc/script_context.cc

```diff
@@ -0,0 +1,172 @@
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "coproc/script_context.h"
+
+#include "coproc/logger.h"
+#include "coproc/pacemaker_context.h"
+#include "coproc/types.h"
+#include "coproc/utils.h"
+#include "model/record_batch_reader.h"
+#include "model/timeout_clock.h"
+#include "vlog.h"
+
+namespace coproc {
+
+script_context::script_context(
+  script_id id, pacemaker_context& ctx, ntp_context_cache&& contexts)
+  : scheduled_task(std::chrono::milliseconds(20))
+  , _ctx(ctx)
+  , _ntp_ctxs(std::move(contexts))
+  , _id(id) {
+    vassert(
+      !_ntp_ctxs.empty(),
+      "Unallowed to create an instance of script_context without having a "
+      "valid subscription list");
+}
+
+ss::future<> script_context::shutdown() {
+    _abort_source.request_abort();
+    return scheduled_task::shutdown().then([this] { _ntp_ctxs.clear(); });
+}
+
+ss::future<> script_context::do_task() {
+    std::vector<process_batch_request::data> requests;
+    requests.reserve(_ntp_ctxs.size());
+    return ss::do_with(
+      std::move(requests),
+      [this](std::vector<process_batch_request::data>& requests) {
+          return ss::parallel_for_each(
+                   _ntp_ctxs,
+                   [this, &requests](const ntp_context_cache::value_type& p) {
+                       return read_ntp(p.second).then(
+                         [&requests](
+                           std::optional<process_batch_request::data> r) {
+                             if (r) {
+                                 requests.push_back(std::move(*r));
+                             }
+                         });
+                   })
+            .then([this, &requests] {
+                if (requests.empty()) {
+                    return ss::now();
+                }
+                return _ctx
+                  .send_request(
+                    process_batch_request{.reqs = std::move(requests)})
+                  .then_wrapped(
+                    [this](ss::future<pacemaker_context::rpc_result> f) {
+                        return process_reply(std::move(f));
+                    });
+            });
+      });
+}
+
+std::optional<storage::log_reader_config>
+script_context::get_reader(const ss::lw_shared_ptr<ntp_context>& ntp_ctx) {
+    auto found = ntp_ctx->offsets.find(_id);
+    vassert(
+      found != ntp_ctx->offsets.end(),
+      "script_id must exist: {} for ntp: {}",
+      _id,
+      ntp_ctx->ntp);
+    const model::offset last_read = found->second;
+    auto stats = ntp_ctx->log.offsets();
+    if (last_read >= stats.dirty_offset()) {
+        return std::nullopt;
+    }
+    return reader_cfg(
+      last_read + model::offset(1), _max_batch_size, _abort_source);
```

Comment:
Its guaranteed to be initialized because when this script_context comes into existance the pacemaker fills the correct ntp_context offsets map with model::offset(0)s.

However I think there is a bug here, looks like we will always start at offset 1, and never 0